### DISCUSSION TITLE
What's a good way to mock window.location?

### DISCUSSION THREAD

**Discussion by uttk-dev** [vitest-dev/vitest#2213] [Q&A] [answered]
I tried mocking `window.location` with `vi.stubGlobal()`, but it seems that the processing of the previous test affects the results of the later test 😭

example code 👇

```js
import { vi } from "vitest"

describe('window.location mock test', () => {
  const hash = '#test-hash';

  // ✅
  test('is passed', () => {
    expect(window.location.hash).not.toBe(hash);
  });


  // ✅
  test('is passed', () => {
    vi.stubGlobal('location', { hash });
    expect(window.location.hash).toBe(hash);
  });

  // ❌
  test('is failed', () => {
    expect(window.location.hash).not.toBe(hash);
  });
});
```

How can I mock `window.location` without affecting other tests?

**Comment by sheremet-va** [comment, marked as answer]
You need to reset the state in `beforeEach` hook

**Reply by uttk-dev** [reply]
Thanks for your comment!
I tried using `vi.restoreAllMocks()` in `beforeEach` or `afterEach` , but the result was the same.
Is there anything I'm doing wrong?

```diff ts
import { vi } from "vitest"

describe('window.location mock test', () => {
  const hash = '#test-hash';

+  beforeEach(() => {
+    vi.restoreAllMocks();
+  });

  // ✅
  test('is passed', () => {
    expect(window.location.hash).not.toBe(hash);
  });


  // ✅
  test('is passed', () => {
    vi.stubGlobal('location', { hash });
    expect(window.location.hash).toBe(hash);
  });

  // ❌
  test('is failed', () => {
    expect(window.location.hash).not.toBe(hash);
  });
});
```

**Reply by sheremet-va** [reply]
Vitest doesn't keep track of your stubs, you need to store the original value, and reset it manually in a hook:

```ts
const original = window.location 
beforeEach(() => {
  window.location = original 
})
```

**Reply by uttk-dev** [reply]
Thank you!

**Reply by sheremet-va** [reply]
There is also a way to do this automatically. Vitest keeps track of your spies, so you can do this:

```ts
vi.spyOn(window, 'location', 'get').mockReturnValue({ hash })
```

If you have `restoreMocks` enabled in config, then you don't need to do anything. Otherwise you need to call `vi.restoreAllMocks` in `beforeEach` hook.

**Reply by Kolobok12309** [reply]
If you use `stubGlobal`, for unstub you need [`unstubAllGlobals`](https://vitest.dev/api/vi.html#vi-unstuballglobals)
And result code
```js
describe('...', () => {
  beforeEach(() => {
    vi.unstubAllGlobals();
  });
});
```

**Reply by kwojcik** [reply]
@sheremet-va What's the best way to do this in browser mode now? `vi.spyOn(window, 'location', 'get')` throws `Cannot redefine property: location`.

**Reply by is-jonreeves** [reply]
@kwojcik is it possible that another test or setup file of yours is already doing something like:
```ts
Object.defineProperty(window, 'location', { value: vi.fn() });
```

If so, you would probably need to avoid it or add the `configurable: true` option:
```ts
Object.defineProperty(window, 'location', { value: vi.fn(), configurable: true });
```
